README.md
# file-parser


This project accepts a file in one of 3 formats: pipe-delimited, comma-delimited and space-delimited.

It expects a header row, and exactly 5 fields: last name, first name, gender, date of birth and favorite color.

Dates should be in the format M/D/YYYY.

## Usage
Call with
```lein run <filename>```

To run with an included sample dataset containing soccer players run:

```
lein run test/data/sample.csv
```

The webserver runs on port 8890.

A list of records may be accessed by doing a get of the /records route.

```
curl localhost:8890/records
```

3 sorts are available:
- Gender `curl localhost:8890/records/gender`
- Birthdate `curl localhost:8890/records/birthdate`
- Last Name (descending) `curl localhost:8890/records/name`

To add new records, send a post to /records with a `person` body parameter containing a comma, pipe,
or space delimited string.

```
curl -XPOST -d "person=Hamm,Mia,Female,Blue,3/17/1972" "http://localhost:8890/records"
```

## License

Copyright © 2019 Rick Hall

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-05-08
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-05-08
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/file-parser/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/file-parser/compare/0.1.0...0.1.1


test/file_parser/api_test.clj
(ns file-parser.api-test
  (:require [clojure.test :refer :all]
            [file-parser.api :as api]
            [file-parser.data :as db]
            [clojure.data.json :as json]
            [io.pedestal.test :as test]
            [io.pedestal.http :as http]))

(defonce server
         (http/start (http/create-server
                       (assoc api/service-map
                         ::http/join? false))))

(def +data-file+ "test/data/sample.csv")

(deftest person-by-name-test
  (testing "getting /records/name returns people ordered by last name descending in valid json"
    (db/make-data +data-file+)
    (is (= (map :LastName (db/get-data [[:LastName :desc]]))
           (map #(get % "LastName")
                (json/read-str
                  (:body
                    (test/response-for
                      (:io.pedestal.http/service-fn server)
                      :get "/records/name"))))))))

(deftest person-by-birthdate-test
  (testing "getting /records/birthdate returns people ordered by date of birth in valid json"
    (db/make-data +data-file+)
    (is (= (map :LastName (db/get-data [:DateOfBirth]))
           (map #(get % "LastName")
                (json/read-str
                  (:body
                    (test/response-for
                      (:io.pedestal.http/service-fn server)
                      :get "/records/birthdate"))))))))

(deftest person-by-gender-test
  (testing "getting /records/gend returns people ordered by gender then last name in valid json"
    (db/make-data +data-file+)
    (is (= (map :LastName (db/get-data [:Gender :LastName]))
           (map #(get % "LastName")
                (json/read-str
                  (:body
                    (test/response-for
                      (:io.pedestal.http/service-fn server)
                      :get "/records/gender"))))))))

(deftest create-person-test
  (testing "Posting to /records creates a person"
    (with-redefs [db/person-data (atom [])]
      (is (= 201 (:status (test/response-for
                            (:io.pedestal.http/service-fn server)
                            :post "/records"
                            :headers {"Content-Type" "application/x-www-form-urlencoded"}
                            :body "person=Hamm Mia Female Blue 3/17/1972"))))
      (is (= 1 (count @db/person-data)))))
  (testing "Posting invalid data returns a 400"
    (is (= 400 (:status (test/response-for
                          (:io.pedestal.http/service-fn server)
                          :post "/records"
                          :headers {"Content-Type" "application/x-www-form-urlencoded"}
                          :body "person=Han Solo, Captain of the Millennium Falcon"))))))

(deftest non-existext-route-test
  (testing "Hitting a route that doesn't exist returns a 404"
    (is (= 404
           (:status
             (test/response-for
               (:io.pedestal.http/service-fn server)
               :get "some/other/route"))))))

test/file_parser/examples.clj
(ns file-parser.examples
  (:require [java-time :as time]
            [clojure.data.csv :as csv]
            [clojure.java.io :as io]))

(def +sample-records+
  [{:LastName "Wambach"
    :FirstName "Abby"
    :Gender "Female"
    :FavoriteColor "Blue"
    :DateOfBirth (time/local-date 1980 6 2)}
   {:LastName "Pogba"
    :FirstName "Paul"
    :Gender "Male"
    :FavoriteColor "Red"
    :DateOfBirth (time/local-date 1993 3 15)}
   {:LastName "Martinez"
    :FirstName "Josef"
    :Gender "Male"
    :FavoriteColor "Peach"
    :DateOfBirth (time/local-date 1993 5 19)}])

(def +example-person-string-map+
  {:LastName "Messi"
   :FirstName "Lionel"
   :Gender "Male"
   :FavoriteColor "Blue"
   :DateOfBirth "6/24/1987"})

(defn map->rows
  [data]
  (let [header (-> data first keys)
        rows (map (apply juxt header) data)
        printable-header (map name header)]
    (reduce conj [printable-header] rows)))

(defn make-csv-file
  "This writes a csv file that may be used in dev and testing"
  ([filename data]
   (with-open [writer (io/writer filename)]
     (csv/write-csv writer (map->rows data))))
  ([data] (make-csv-file "sample.csv" data))
  ([] (make-csv-file +sample-records+)))


test/file_parser/core_test.clj
(ns file-parser.core-test
  (:require [clojure.test :refer :all]
            [file-parser.core :refer :all]
            [file-parser.api :as api]
            [file-parser.examples :refer [+sample-records+]]))

(deftest main-only-works-with-one-arg-test
  (with-redefs [file-parser.core/produce-output (constantly true)
                api/start (constantly true)
                file-parser.data/make-data (constantly true)]
    (testing "The main function calls csv-file->map when it has 1 argument"
      (is (true? (-main "test"))))
    (testing "The main function does not call csv-file->map when it has no arguments"
      (is (nil? (-main))))
    (testing "The main function does not call csv-file->map when it has more than 1 argument"
      (is (nil? (-main "one" "two" "three"))))))


src/file_parser/core.clj
(ns file-parser.core
  (:require [file-parser.sorter :refer [sort-by-keys]]
            [file-parser.person :as person]
            [file-parser.data :as db]
            [file-parser.api :as api]))

(defn print-table
  [table]
  (doseq [line table]
    (println line)))

(defn produce-output
  []
  (let [sorts [[:Gender :LastName]
               [:DateOfBirth :LastName]
               [:LastName :desc]]]
    (doseq [sort sorts]
      (-> sort
          db/get-data
          person/format-for-printing
          print-table))))

(defn -main
  [& args]
  (if
    (= (count args) 1)
    (let [data-load-result (db/make-data (first args))]
      (if (= (type data-load-result) java.lang.String)
        (println "Loading data failed: " data-load-result)
        (do
          (produce-output)
          (api/start))))
    (println "Please pass a single argument, containing the filename you wish to load")))


src/file_parser/data.clj
(ns file-parser.data
  (:require [file-parser.sorter :refer [sort-by-keys]]
            [clojure.string :as str]
            [file-parser.person :as person]))

(def person-data (atom []))

(defn get-data
  ([sort]
   (let [data @person-data]
     (if sort
       (sort-by-keys data sort)
       data)))
  ([] (get-data nil)))

(defn add-person
  [new-person]
  (swap! person-data conj new-person))

(defn read-lines
  [filename]
  (let [x (slurp filename)
        lines (str/split-lines x)]
    lines))

(defn find-delimiter
  "Look for a supported delimiter char in a string and return a regex pattern"
  [sample-string]
  (cond
    (str/includes? sample-string "|") #"\|"
    (str/includes? sample-string ",") #","
    (str/includes? sample-string " ") #"\s"))

(defn line->Person
  ([line delimiter]
   (try
     (-> (str/split line delimiter)
         (person/strings->Person))
     (catch Throwable t
       "Error processing line")))
  ([line]
    (line->Person line (find-delimiter line))))

(defn delimited-strings->Persons
  [[_header & rows]]
  (let [delimiter (find-delimiter (first rows))
        results (map #(line->Person % delimiter) rows)]
    (if (some #(= java.lang.String (type %)) results)
      "Error processing file"
      results)))

(defn csv-file->Persons
  [filename]
  (try
    (-> (read-lines filename)
        (delimited-strings->Persons))
    (catch java.io.FileNotFoundException e
      "File not found")
    (catch Throwable t
      (str "An error occurred" (type t)))))

(defn make-data
  [filename]
  (let [data (csv-file->Persons filename)]
    (if (= (type data) java.lang.String)
      data
      (reset! person-data data))))


src/file_parser/person.clj
(ns file-parser.person
  (:require [java-time :as time]
            [clojure.string :as str]))

(def +output-fields+ [:LastName :FirstName :Gender :FavoriteColor :DateOfBirth])

(defrecord Person [LastName FirstName Gender FavoriteColor DateOfBirth])

(defn dob-string->date
  "Translate DateOfBirth to a date type"
  [row]
  (try
    (if (:DateOfBirth row)
      (update row :DateOfBirth (fn [s]
                                 (when s (time/local-date "M/d/yyyy" s))))
      row)
    (catch Throwable t
      "Error parsing date")))

(defn dob-date->string
  [row]
  (if (:DateOfBirth row)
    (update row :DateOfBirth (fn [d] (time/format "M/d/yyyy" d)))
    row))

(defn strings->Person
  [strings]
  {:pre [(= 5 (count strings))]}
  (try
    (let [person-map (->> strings
                          (zipmap +output-fields+)
                          (dob-string->date))]
      (if (= java.lang.String (type person-map))
        person-map
        (map->Person person-map)))
    (catch Throwable t "Error creating person from strings")))

(defn map-of-strings->Person
  [map-of-strings]
  (-> map-of-strings
      (dob-string->date)
      (map->Person)))

(defn Person->map-of-strings
  [person]
  (-> person
      (select-keys +output-fields+)
      (dob-date->string)))

(defn format-for-printing
  [data]
  (let [header (map name +output-fields+)]
    (->> data
         (map dob-date->string)
         (map (apply juxt +output-fields+))
         (reduce conj [header])
         (map #(str/join "\t" %)))))

(defn Person->json
  [person]
  (str
    (->> person
         (Person->map-of-strings)
         (reduce-kv
           (fn [m k v] (str m (pr-str (name k)) ":" (pr-str v)))
           "{"))
    "}"))


src/file_parser/sorter.clj
(ns file-parser.sorter)

(defn at-bottom-level?
  [data]
  (or (map? (first data))
      (empty? (first data))))

(defn reassemble-data
  "This flattens a sequence of sequence of maps to a sequence of maps, recursively"
  [data]
  (if (at-bottom-level? data)
    data
    (recur (apply concat data))))

(defn sort-asc-or-desc
  [key data]
  (let [sort-fn (if (vector? key)
                  (first key)
                  key)
        reverse-fn (if (and
                         (vector? key)
                         (= :desc (first (rest key))))
                     reverse
                     identity)]
    (->> data
         (sort-by sort-fn)
         (reverse-fn)
         (partition-by sort-fn))))

(defn- sort*
  "This does the actual work of sorting, but returns a nested data-structure that isn't useful on its own.
  Call sort-by-keys for sorting"
  [data remaining-keys]
  (cond
    (= 0 (count remaining-keys)) data
    (not (at-bottom-level? data)) (map #(sort* % remaining-keys) data)
    :else (recur (sort-asc-or-desc (first remaining-keys) data) (rest remaining-keys))))

(defn sort-by-keys
  [data sort-keys]
  (let [sort-keys (take 5 sort-keys)]                        ; only supporting sorting to 5 levels for safety
    (reassemble-data (sort* data sort-keys))))


src/file_parser/api.clj
(ns file-parser.api
  (:require [io.pedestal.http :as http]
            [io.pedestal.http.route :as route]
            [io.pedestal.http.body-params :as body-params]
            [file-parser.data :as db]
            [file-parser.person :as person]))

(defn ok
  [body & {:as headers}]
  {:status 200 :body body
   :headers (assoc headers "Content-Type" "application/json")})

(defn created
  [body & {:as headers}]
  {:status 201 :body body
   :headers headers})

(defn error
  [body & {:as headers}]
  {:status 400 :body body
   :headers headers})

(defn people->json
  [people]
  (str
    (->> people
         (map person/Person->json)
         (reduce str "["))
    "]"))

(def person-list
  {:name :person-list
   :enter
         (fn [context]
           (assoc context :response (ok (people->json (db/get-data)))))})

(def person-by-gender
  {:name :person-by-gender
   :enter
         (fn [context]
           (assoc context :response (ok (people->json (db/get-data [:Gender :LastName])))))})

(def person-by-birthdate
  {:name :person-by-birthdate
   :enter
         (fn [context]
           (assoc context :response (ok (people->json (db/get-data [:DateOfBirth])))))})

(def person-by-name
  {:name :person-by-name
   :enter
         (fn [context]
           (assoc context :response (ok (people->json (db/get-data [[:LastName :desc]])))))})

(def person-create
  {:name :person-create
   :enter (fn [context]
            (let [new-person (get-in context [:request :form-params :person])
                  person (db/line->Person new-person)]
              (if (= java.lang.String (type person))
                (assoc context :response (error "Failure parsing person data"))
                (do (db/add-person person)
                    (assoc context :response (created "Person added"))))))})

(def routes
  (route/expand-routes
    #{["/records" :post [(body-params/body-params) http/html-body person-create]]
      ["/records" :get person-list :route-name :person-list]
      ["/records/gender" :get person-by-gender :route-name :person-by-gender]
      ["/records/birthdate" :get person-by-birthdate :route-name :person-by-birthdate]
      ["/records/name" :get person-by-name :route-name :person-by-name]}))

(def service-map
  {::http/routes routes
   ::http/type   :jetty
   ::http/port   8890})

(defn start []
  (http/start (http/create-server service-map)))
